### ISSUE TITLE
IO.delay.flatMap is not cancelable in 1.0.0

### ISSUE THREAD

**Issue by maphi** [typelevel/cats-effect#363]
Just curious if this is the intended behaviour. Minimal Example: 

```
import cats.effect._
import cats.implicits._
import scala.concurrent.duration._
import scala.concurrent.ExecutionContext.Implicits.global

implicit val cs = IO.contextShift(global)
implicit val timer = IO.timer(global)

val prog = for {
  _ <- IO{Thread.sleep(1000); println("Finished 1")}
  _ <- IO{Thread.sleep(1000); println("Finished 2")}
  _ <- IO{Thread.sleep(1000); println("Finished 3")}
} yield ()

scala>   prog.timeout(1500.millis).unsafeRunAsync(println)

scala> Finished 1
Left(java.util.concurrent.TimeoutException: 1500 milliseconds)
Finished 2
Finished 3
```

**Comment by SystemFw** [comment]
You can 't cancel a `delay` block, the JVM doesn't really give you tools to do that. You can cancel the equivalent behaviour (which also doesn't block threads), by replacing `Thread.sleep` with `timer.sleep(1.second) *> IO(println(...))`.

As for canceling the flatMap chain instead, the cancelation flag it's checked at async boundaries (so `sleep`, waiting on a `Deferred`, `shift`, and generally anything built in terms of `async`, `asyncF`, `start` or `cancelable`)

**Comment by maphi** [comment]
Ok i understand. But why is cancellation only possible at async boundaries? If for example i do some expensive calculations inside the IO.delay and chain them, why shouldn't it be possible to interrupt the cain at the flatMap call (without adding an async boundary)?

**Comment by alexandru** [comment]
@maphi it's expensive to check the cancellation flag at each flatMap. Usually in `delay` the computations are cheap. And also cancellation being concurrent, without synchronization (via your own logic in `IO.cancelable`) the thread doing the loop can notice that cancellation flag with a significant delay anyway. So it's not worth it.

Also note that this is not cancelable:

```scala
IO{Thread.sleep(1000); println("Finished 1")}
```

Meaning that using the constraints we have, there's no way to make this `IO` not print `Finished` at the end on evaluation. So your options would be:

1. describe this via `IO.cancelable` 
2. use `IO.cancelBoundary`

This also works:

```scala
IO(Thread.sleep(1000)) *> IO.cancelBoundary *> IO(println("Finished 1"))
```

**Comment by alexandru** [comment]
Note this is described in our docs, make sure to check that out: https://typelevel.org/cats-effect/datatypes/io.html#concurrency-and-cancellation

**Comment by maphi** [comment]
Ah just missed that point in the docs. Thanks for clarifying!
